bin/weewx/drivers/ws1.py

```diff
@@ -221,6 +203,209 @@ def _decode(s, multiplier=None, neg=False):
+               (self.conn_info[0], self.conn_info[1]))
+        try:
+            self.net_socket.close()
+        except (socket.error, socket.herror, socket.timeout), ex:
+            logerr("Cannot close connection to %s:%d for some reason: %s" % (
+                self.conn_info[0], self.conn_info[1], ex))
+            raise weewx.WeeWxIOError(ex)
+
+    def get_readings(self):
+        if self.rec_start is not True:
+            # Find the record start
+            if DEBUG_READ: loginf("Attempting to find record start..")
+            buf = ''
+            while True:
+                try:
+                    buf += self.net_socket.recv(8, socket.MSG_WAITALL)
+                except (socket.error, socket.timeout), ex:
+                    raise weewx.WeeWxIOError(ex)
+                if DEBUG_READ: loginf("(searching...) buf: %s" % buf)
+                if '!!' in buf:
+                    self.rec_start = True
+                    if DEBUG_READ: loginf("Record start found!")
+                    # Cut to the record start
+                    buf = buf[buf.find('!!'):]
+                    if DEBUG_READ: loginf("(found!) buf: %s" % buf)
+                    break
+            # Add the rest of the record
+            try:
+                buf += self.net_socket.recv(
+                    PACKET_SIZE - len(buf), socket.MSG_WAITALL)
+            except (socket.error, socket.timeout), ex:
+                raise weewx.WeeWxIOError(ex)
+        else:
+            # Keep receiving data until we find an exclamation point or two
+            try:
+                buf = self.net_socket.recv(2, socket.MSG_WAITALL)
+            except (socket.error, socket.timeout), ex:
+                raise weewx.WeeWxIOError(ex)
+            while True:
+                if buf == '\r\n':
+                    # CRLF is expected
+                    if DEBUG_READ: loginf("buf is CRLF")
+                    buf = ''
+                    break
+                elif '!' in buf:
+                    excmks = buf.count('!')
+                    # Assuming exclamation points are at the end of the buffer
+                    buf = buf[len(buf) - excmks:]
+                    if DEBUG_READ: loginf("buf has %d exclamation points." % (
+                        excmks))
+                    break
+                else:
+                    try:
+                        buf = self.net_socket.recv(2, socket.MSG_WAITALL)
+                    except (socket.error, socket.timeout), ex:
+                        raise weewx.WeeWxIOError(ex)
+                    if DEBUG_READ: loginf("buf: %s" % ' '.join(
+                        ['%02X' % ord(bc) for bc in buf]))
+            try:
+                buf += self.net_socket.recv(
+                    PACKET_SIZE - len(buf), socket.MSG_WAITALL)
+            except (socket.error, socket.timeout), ex:
+                raise weewx.WeeWxIOError(ex)
+        if DEBUG_READ: loginf("buf: %s" % buf)
```

Comment:
any DEBUG_READ messages should be logdbg, not loginf.  also, prolly safer to keep it on two lines instead of inlining a conditional.